crates/nu-plugin-engine/src/declaration.rs

```diff
@@ -123,4 +123,40 @@ impl Command for PluginDeclaration {
     fn plugin_identity(&self) -> Option<&PluginIdentity> {
         Some(&self.source.identity)
     }
+
+    fn get_dynamic_completion(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        arg_type: &ArgType,
+    ) -> Result<Option<Vec<String>>, ShellError> {
```

Comment:
I think the error is discarded at the call site. And  I don't see the necessity of Option here, would you please document the return type in more details?

Reply:
Also, I'm in favor of `Vec<SemanticSuggestion>` for better customizability.

Reply:
I agree about letting plugins return `Vec<SemanticSuggestion>`s, no need to restrict plugins to strings.

But wrapping the suggestions in an `Option` can be useful to differentiate between "there are no suggestions" (`Some(vec![])`) and "I couldn't find any suggestions, please fall back to default completions" (`None`)

Reply:
Ah, didn't realize that. I think moving `SemanticSuggestion` and `SuggestionKind` to nu-protocol is fine, but might want to ask someone knowledgeable about nu-protocol.

Another option is returning `Vec<Suggestion>` and always using `SuggestionKind::Value`. But that makes it less customizable, especially if we decide to add a new field to `SemanticSuggestion` in the future. @blindFS thoughts?

Reply:
Basically if we use `Vec<SemanticSuggestion>` as the return type, we want the matching done here in this function.

But it could be too wordy for each command to take care of those prefix/span/completion_option stuff.

We probably need new api for this, shall we leave it for future PRs with a TODO?

Reply:
Hmm, yeah we should probably provide a way for commands to make use of `NuMatcher` if we're going to have them return a filtered `Vec<SemanticSuggestion>`. I think it's okay to make it return a `Vec<SemanticSuggestion>` for now, move `SemanticSuggestion` into nu-protocol, and then leave a TODO to figure out where to move `NuMatcher` later. I'd rather just move `NuMatcher` into nu-protocol than create a new crate, since the added nucleo-matcher dependency isn't too big.

Reply:
Let's stick to vec of strings then, I'll see what I can do later when trying to call DotNuCompletion in `use/source`, CommandCompletion in `which`, etc.

Reply:
Is it safe to change the public API? I don't want people writing completions for their plugins that return `Vec<String>` and then finding out that we've changed it to `Vec<???>` in the next release.

Can we instead make our own ripoff of `Suggestion` in `nu-protocol` and have Commands return that? In `nu-cli`, we could manually map these `RipoffSuggestion`s to `SemanticSuggestion`s. That way, we wouldn't have to change the public API.

Reply:
Does it make sense to put NuMatcher to `nu-utils`? I mean the generic version of it.

Reply:
That sounds fine to me. It depends on nucleo-matcher, but that's not a huge dependency. Makes more sense than putting it in nu-protocol.

Reply:
Sorry I'm not familiar in this area.  If so, do we need to pass span, and maybe offset to plugin command?

Reply:
Actually I don't think we need span for internal/plugin commands. We don't flatten the input like for external commands. The completer/matcher only gets prefix of current active argument.

Reply:
Hmm, I just realized that `get_dynamic_completion` doesn't get the buffer. Sorry to keep holding up this PR, but I think it'd be better if it also accepted the buffer/text typed by the user.

We don't need to pass in a span. Later, we can give plugins the ability to return spans, and then we would translate the span to a Reedline span with the right offset.